### ISSUE TITLE
Support parsing vendor-specific interval literals

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#7445]
Currently, `YearToMonth` and `DayToSecond` support parsing SQL standard interval literals (which can also be used in the parser, see #7518). It would be great if other formats could be parsed as well, including (from the [PostgreSQL docs](https://www.postgresql.org/docs/current/static/datatype-datetime.html)):

Style Specification | Year-Month Interval | Day-Time Interval | Mixed Interval
-- | -- | -- | --
sql_standard | 1-2 | 3 4:05:06 | -1-2 +3 -4:05:06
postgres | 1 year 2 mons | 3 days 04:05:06 | -1 year -2 mons +3 days -04:05:06
postgres_verbose | @ 1 year 2 mons | @ 3 days 4 hours 5 mins 6 secs | @ 1 year 2 mons -3 days 4 hours 5 mins 6 secs ago

Setting aside the mixed interval for the moment (as we don't have a type to represent that)

**Comment by ciaccia** [comment]
Minor change to the binder to support also plus signs other than minuses. I was not able to reproduce them but according to postgresql's source code they could appear as well.

-> [PostgresIntervalBinder-fixed.txt](https://github.com/jOOQ/jOOQ/files/2614561/PostgresIntervalBinder-fixed.txt)

**Comment by lukaseder** [comment]
Thanks a lot, @ciaccia. I think I'd like to avoid using regular expressions for 2 reasons:

1. They always appear in benchmarks, because of the `Matcher`'s internal `int[]` that is needlessly allocated at every execution. When parsing millions of intervals, this will definitely hurt
2. The accepted syntaxes in PostgreSQL are quite funky. It is extremely lenient. E.g. this is valid: `SELECT INTERVAL '1 sec 1 y'`

I guess an actual parser will be needed. It will also be used by the `org.jooq.Parser` itself.

**Comment by lukaseder** [comment]
> I see your points about avoiding RE (I guess you are referring to #7986 here)

For the avoidance of doubt: That issue was not caused by regular expressions but hashmaps, but yes, it does show how critical `Binding` code is when large result sets are involved.

> My binder was only meant to parse data (in intervalstyle = postgres format) from the JDCB getString() method, was not meant to parse "input" literals... (input = entered by the client, not produced by the server).

There can be only one implementation for any purpose :) Also, I've seen PostgreSQL break backwards compatibility in the past in subtle ways. Better play safe here. Since I'm planning to actually parse the literal using a parser rather than a regex, it won't be much more complicated to allow for additional syntax.

> Please excuse me for asking, but why would a binder be able to parse client-side Postgresl's literals? 

Not the binder itself, maybe, but as I said, the parser will definitely need to do that. E.g.:

```java
DSL.using(config)
   .parser()
   .parse("SELECT INTERVAL '1 y 1 sec'");
```

And since there will be only one implementation, it will need to be able to parse any syntax that is supported by the server.

**Comment by lukaseder** [comment]
I do wonder, however, if a user should really use the `Duration` type to map to a PostgreSQL `INTERVAL`. A `Duration` corresponds to a `DAY TO SECOND` interval, but cannot represent a `YEAR TO MONTH` or "mixed" interval in PostgreSQL.

**Comment by ciaccia** [comment]
Hi Lukas,
Why are you saying that Duration corresponds to a `DAY TO SECOND` interval? According to  [official Java Tutorials](https://docs.oracle.com/javase/tutorial/datetime/iso/period.html):

>  A Duration measures an amount of time using time-based values (seconds, nanoseconds)

On Java side the duration is not tied to days, it can span years as well. I guess the real question is the following. Consider a Postgresql statement where the have two intervals spanning both 86400 seconds (column 1 and 2), in column 3 we subtract them and column 4 extracts the seconds of column 3:

```sql
select '1 day'::interval, '86400 seconds'::interval, ('1 day'::interval - '86400 seconds'::interval),
   EXTRACT(EPOCH FROM ('1 day'::interval - '86400 seconds'::interval));
```
The query returns:
> "1 day"	"24:00:00"	"1 day -24:00:00"	"0"

Is it acceptable that columns 1+2 and columns 3+4 in Java have the exact same value despite being represented in different ways in Postgres? If the answer is yes, Duration is very likely the way to go. Otherwise a new datastructure is needed to expose Postgres internal representation to the Java user.

**Comment by lukaseder** [comment]
Aaah, the wonderfully subtle world of date time arithmetic :)

So, you're saying a day has 86400 seconds, right? What about daylight savings time? There are days with 23h or 25h. Or leap seconds. But the `Duration` doesn't care about that. The `Duration` corresponds to the SQL standard `DAY TO SECOND` interval type, which is based on seconds and fractional seconds (day being just a convenient way to say 24h or 86400 seconds).

Try extracting the epoch from a `YEAR TO MONTH` interval. It will get much more curious. Now, you're extracting the epoch and thus "destroying" some interval information. Because these two are not the same "durations":

```sql
SELECT date '2004-01-01' + interval '1 year'; -- 366 days
SELECT date '2005-01-01' + interval '1 year'; -- 365 days
SELECT EXTRACT (EPOCH FROM INTERVAL '1 year') / 86400 ; -- 365.25 days, a compromise that isn't very accurate
```

jOOQ has the `org.jooq.types.Interval` types that match the SQL definition of intervals. It is not exactly the same as the JSR 310 definition. It's also a compromise. PostgreSQL went down the rabbit hole and allowed for "mixed intervals" where `YEAR TO MONTH` and `DAY TO SECOND` interval types can be mixed. This further complicates things.

As you can see, some intervals can be represented as a `java.time.Duration`, but not all.

**Comment by ciaccia** [comment]
Hi @lukas,
Thank you for your example, I didn't realize that two intervals could behave differently when added or subtracted to a date despite having the same number of seconds when considered on their own (such as java Duration). This example also explains why you can write `strange` intervals:
```sql
SELECT date '2004-01-01' + interval '1 year -365 days'; -- 2004-01-02
SELECT date '2005-01-01' + interval '1 year -365 days'; -- 2005-01-01
```
Probably here jOOQ should map mixed-intervals to a Duration+Period combo, which makes everything more complex. I am not sure anymore what the best approach would be in this case...

**Comment by lukaseder** [comment]
My plan is to create a new `YearToSecond` interval that is composed of the existing `YearToMonth` and `DayToSecond` intervals: #8077. This needs a lot more study, though.

I'm not going to use the JSR-310 types here because their semantics is different from SQL's. This doesn't mean that they're wrong, they just implement a different strategy.

**Comment by lukaseder** [comment]
`YearToSecond` intervals (#8077) are now supported in jOOQ 3.12. Verbose intervals for PostgreSQL are also supported through #7518 and https://github.com/jOOQ/jOOQ/commit/126c8b496e432fac6af319fb84d1c05751401b33. Probably some syntax is still not available. We'll fix that as we get feature requests...

**Comment by ciaccia** [comment]
Hi @lukaseder,
Thanks a lot for the nice addition! Apart from the new parser, will jOOQ 3.12 automatically use the new types in the code generator? If I have a column whose type is `inteval`, will the generated class declare its type to be `YearToSecond`?

**Comment by lukaseder** [comment]
@ciaccia : That's a separate issue: #2230

**Comment by ciaccia** [comment]
Hi @lukaseder,
I had the chance to look at your parser (but not run it), I am not sure I understand it completely. Is your parser actually covering all the possible formats for Postgresql's mixed intervals? What happens if I enter a "hour:minute:second.fraction" string? Looking at your code I don't understand how it can parse the following literals:
```sql
SELECT '3 years 1.5 months 246:23:12.567'::interval;
SELECT '246:23:12.567 3 years 1.5 months'::interval;
```

Probably I am missing something here but I just wanted to let you know.

**Comment by lukaseder** [comment]
> I had the chance to look at your parser (but not run it), I am not sure I understand it completely. Is your parser actually covering all the possible formats for Postgresql's mixed intervals?

Probably not. Thus my comment *"Probably some syntax is still not available. We'll fix that as we get feature requests..."*

> What happens if I enter a "hour:minute:second.fraction" string?

Indeed, I had omitted parsing the mixed syntax for now, as I found it not really clearly specified. Thanks for digging into the PostgreSQL sources. That'll definitely help. I will create a new issue from your comments: #8271

Please feel free to add additional syntax you may have discovered there.

**Comment by ciaccia** [comment]
Hi @lukaseder,
I noticed the part about not having covered all the possible syntaxes, but the syntax with the _time-like_ part is the one that the server outputs per default. So in case you want to use the parser both for the parsing API and the binder, then you should probably support it as well.

For example the following query outputs `3 days 02:01:31`.

```sql
CREATE TABLE interval_test (
   id bigserial PRIMARY KEY,
   duration interval
);

insert into interval_test("duration") values ('3 days 2 hours 1 minute 31 seconds');
SELECT * FROM interval_test;
```

**Comment by lukaseder** [comment]
That's not what I'm getting. In DBeaver, I'm getting:

```
id |duration                                        |
---|------------------------------------------------|
1  |0 years 0 mons 3 days 2 hours 1 mins 31.00 secs |
```

**Comment by lukaseder** [comment]
Interesting. I'm not aware of tampering with this setting. Just clicking next next next on the installer :) Good to know!

**Comment by Artus2b** [comment]
Hi Lukas,
we actually implemented a `java.time.Duration` <> `PGInterval` Binding+Converter before this version.
So this new feature allows us to just use a Converter `java.time.Duration` <> `YearToSecond`. Which is made very easy, as the mapping methods are in `YearToSecond`.
Thanks! :+1: 

However we took different choice in the conversion of a `year`: 360 days, instead of 365.25 when using jOOQ's `YearToSecond`, based on this test:
```SQL
SELECT justify_interval(interval '390 days');
1 years 1 mons 0 days 0 hours 0 mins 0.00 secs

SELECT interval '1 year' = interval '360 days';
true
```

So there is a different conversion rate from `interval` to `interval` than by converting to a number of seconds (using `EXTRACT(EPOCH FROM ...)`)..
How is that possible?
```SQL
select justify_interval(interval '1 second' * EXTRACT(EPOCH FROM interval '1 year'));
1 year 5 days 06:00:00
```
